runners/flink/1.15/src/main/java/org/apache/beam/runners/flink/transform/sql/FlinkSql.java

```diff
@@ -0,0 +1,463 @@
+    validateOutputTableName(name);
+    return Preconditions.checkStateNotNull(
+        (TableInfo.AdditionalOutputTableInfo<?>) outputTableInfo.get(name));
+  }
+
+  public boolean hasSpecifiedMainOutputTableName() {
+    return !DEFAULT_MAIN_OUTPUT_TABLE_NAME.equals(mainOutputTableName);
+  }
+
+  public boolean isMainInputUsed() {
+    return !UNUSED_MAIN_INPUT_TABLE_NAME.equals(mainInputTableName);
+  }
+
+  // ------------------ private helper methods ---------------------------
+
+  private void validateInputTableName(String name) {
+    if (!inputTableInfo.containsKey(name)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "The input table %s is not found. The valid table names for input are %s",
+              name, inputTableInfo.keySet()));
+    }
+  }
+
+  private void validateOutputTableName(String name) {
+    if (!queries.containsKey(name)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "The output table %s is not found. The valid table names for output are %s",
+              name, queries.keySet()));
+    }
+  }
+
+  // ------------------ public nested classes -------------------------
+
+  /** Registers Flink SQL PTransform URN. */
+  @AutoService(TransformPayloadTranslatorRegistrar.class)
+  @SuppressWarnings("rawtypes")
+  public static class FlinkTransformsRegistrar implements TransformPayloadTranslatorRegistrar {
+    @Override
+    public Map<
+            ? extends Class<? extends PTransform>,
+            ? extends PTransformTranslation.TransformPayloadTranslator>
+        getTransformPayloadTranslators() {
+      return ImmutableMap
+          .<Class<? extends PTransform>, PTransformTranslation.TransformPayloadTranslator>builder()
+          .put(
+              MultiOutputSqlTransform.class,
+              PTransformTranslation.TransformPayloadTranslator.NotSerializable.forUrn(
+                  FLINK_SQL_URN))
+          .build();
+    }
+  }
+
+  /** Registers Flink SQL PTransform to the Flink runner. */
+  @AutoService(FlinkCustomTransformTranslatorRegistrar.class)
+  public static class FlinkSqlTransformsRegistrar
+      implements FlinkCustomTransformTranslatorRegistrar {
+    @Override
+    public Map<String, FlinkStreamingPipelineTranslator.StreamTransformTranslator<?>>
+        getTransformTranslators() {
+      return ImmutableMap
+          .<String, FlinkStreamingPipelineTranslator.StreamTransformTranslator<?>>builder()
+          .put(FLINK_SQL_URN, new FlinkSQLTransformTranslator<>())
+          .build();
+    }
+  }
+
+  /**
+   * Translator class for {@link FlinkSql}. The input Tables for SQL are available in two ways. 1.
+   * From the upstream PTransforms. In this case, the table name is
+   *
+   * @param <InputT> the input data type.
+   * @param <OutputT> the output data type.
+   */
+  private static class FlinkSQLTransformTranslator<InputT, OutputT>
```

Comment:
Makes sense. In beam typically the transform and translator are separated due to the transform is engine-agnostic. Since this transform is for Flink only, I guess making it in the same package doesn't matter too much.